src/main/java/org/simplify4u/plugins/ArtifactResolver.java

```diff
@@ -0,0 +1,262 @@
+     * @param projectResolvedArtifacts a separate set of artifacts with versions as resolved by dependency resolution.
+     *                                 Input is expected to have significant overlap, but may contain deviations such as
+     *                                 missing artifacts.
+     * @return Returns set of artifacts with versions updated according to provided input set.
+     * @throws MojoExecutionException In case of failure to resolve artifact both through dependency resolution process
+     * and manual resolving.
+     */
+    private Set<Artifact> updateArtifactResolvedVersions(final Iterable<Artifact> allArtifacts,
+            final Iterable<Artifact> projectResolvedArtifacts) throws MojoExecutionException {
+        final LinkedHashSet<Artifact> result = new LinkedHashSet<>();
+        for (final Artifact artifact : allArtifacts) {
+            final Optional<Artifact> projectResolved = stream(projectResolvedArtifacts.spliterator(), false)
+                    .filter(s -> s.getArtifactId().equals(artifact.getArtifactId())
+                            && s.getGroupId().equals(artifact.getGroupId()))
+                    .findFirst();
+            if (projectResolved.isPresent()) {
+                // add artifact with version as resolved by Maven dependency resolution
+                artifact.setVersion(projectResolved.get().getVersion());
+                result.add(resolve(artifact));
+            } else if (artifact.isResolved()) {
+                // add artifact with listed version
+                result.add(artifact);
+            } else {
+                // failed to resolve artifact with definite version
+                throw new MojoExecutionException("Failed to determine definite version for artifact " + artifact);
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Retrieves the PGP signature file that corresponds to the given Maven artifact.
+     *
+     * @param   artifacts
+     *          The artifacts for which a signatures are desired.
+     * @return  Either a Maven artifact for the signature file, or {@code null} if the signature
+     *          file could not be retrieved.
+     *
+     * @throws MojoExecutionException
+     *          If the signature could not be retrieved and the Mojo has been configured to fail
+     *          on a missing signature.
+     */
+    Map<Artifact, Artifact> resolveSignatures(final Iterable<Artifact> artifacts, final SignatureRequirement requirement) throws MojoExecutionException {
+        log.debug("Start resolving ASC files");
+
+        final LinkedHashMap<Artifact, Artifact> artifactToAsc = new LinkedHashMap<>();
+        for (Artifact artifact : artifacts) {
+            final Artifact ascArtifact = resolveSignature(artifact, requirement);
+
+            if (ascArtifact != null || requirement == SignatureRequirement.STRICT) {
+                artifactToAsc.put(artifact, ascArtifact);
+            }
+        }
```

Comment:
maybe java8 strem will be ok

Reply:
Are there any concerns?

Reply:
Ah, right! I misunderstood your comment. I'll have a look, see if it makes sense. It should be okay.